online_meetings/report.py

```diff
@@ -0,0 +1,159 @@
+
+
+def zoom_loop(url: str, headers: Dict[str, Union[str, int]], json_attribute_name: str,
+              params: Dict[str, Union[str, int]], page_token: bool = False) -> list:
+    # Need a fresh copy of dicts
+    total_list = []    # get total page count
+    total_page_count = get_total_page_count(url, headers, params)
+    logger.info(f"Total page number {total_page_count}")
+    # Either go by the page number or token
+    while (page_token or params.get('page_number') <= total_page_count):
+        if (params.get("page_number")):
+            logger.info(f"Page Number: {params.get('page_number')} out of total page number {total_page_count}")
+
+        response = requests.request(
+            "GET", url, headers=headers, params=params)
+        status_code = response.status_code
+        # Rate limited, wait a few seconds
+        if status_code == 429:
+            # This is what the header should be
+            retry_after = response.headers.get("Retry-After")
+            sleep_time = DEFAULT_SLEEP_TIME
+            if retry_after and retry_after.isdigit():
+                logger.warning(f"Received status 429, need to wait for {retry_after}")
+                sleep_time = retry_after
+            else:
+                logger.warning(f"No Retry-After header, setting sleep loop for {DEFAULT_SLEEP_TIME} seconds")
+            while status_code == 429:
+                time.sleep(sleep_time)
+                response = requests.request(
+                    "GET", url, headers=headers, params=params)
+                status_code = response.status_code
+
+        if status_code != 200:
+            logger.warning(f'Received irregular status code: {status_code}')
+            break
+        else:
+            try:
+                results = json.loads(response.text.encode('utf8'))
+
+                total_list.extend(results[json_attribute_name])
+                logger.info(f'Current size of list: {len(total_list)}')
+
+                # go retrieve next page
+                if results.get("next_page_token"):
+                    page_token = results.get("next_page_token")
+                    params["next_page_token"] = page_token
+                elif params.get("page_number"):
+                    params["page_number"] += 1
+                else:
+                    logger.info("No more tokens and not paged!")
+                    break
+            except json.JSONDecodeError:
+                logger.exception('JSONDecodeError encountered')
+                logger.info('No more pages!')
+                break
+    return total_list
+
+
+# run users report
+run_report('/v2/users', headers, 'users', {"status": "active", "page_number": 1}, page_token=False, use_date=False)
+# run meetings report
+run_report('/v2/metrics/meetings', headers, 'meetings', {"type": "past"}, page_token=True, use_date=True)
+# run webinars report
+run_report('/v2/metrics/webinars', headers, 'webinars', {"type": "past"}, page_token=True, use_date=True)
```

Comment:
Can you time these all call so that we know how long it ran

Reply:
I think timing everything can come as part of a future fix. I'd like to have a decorator or class extension we can use for both the job run and this timing. I'll create this as a new issue.